### ISSUE TITLE
Request: optimize decode method

### ISSUE THREAD

**Issue by nonujigu** [bit-dream/can-dbc#2]
I tested the decode method with a Raspberry Pi 3 running Node-RED with a dbc file with 897 different can messages. It works fine, but the performance could be better. The decode method needed mostly 2 milliseconds, sometimes just 1 millisecond. If there are 500 CAN messages per second then realtime decoding gets difficult.
The potential bottleneck could be that the resulting object from the parsed dbc file utilizes a javascript map for the CAN messages where the message names are the keys. The getMessageById method loops through this map until it finds the CAN ID. If the decode method could access a map where the keys are the CAN IDs then this could potentially boost the decode performance.

I am aware that this library is more targeted towards manipulating dbc files. However, I think it might be worth to take a look at this.

By the way: Parsing the dbc file with 897 different can messages on a Raspberry Pi 3 took more then 12 seconds.

**Comment by bit-dream** [comment]
nonujigu, great suggestions. I like the idea of migrating the getMessageById call to using a Map with the IDs being the keys in the decode function. That would hopefully optimize the lookup time of the function and possibly help with timing when bus utilization is high. Should be simple enough to get those changes in and get a pull request going.

Additionally, I'll take a look into further optimizations around parsing of the dbc file. 12 seconds seems excessive, so I'm sure there are things that can be improved upon.

**Comment by nonujigu** [comment]
I have a few thoughts on this:

### Maps vs Objects
As pointed out in [this article](https://www.zhenghao.io/posts/object-vs-map), utilizing integers as keys for a map could even lead to a performance disadvantage compared to a simple object. So for the decode scenario, a simple object that utilizes stringified integers as keys could even be the best choice. However, it is probably necessary to implement both variants and compare the results to be sure. Also, the results could heavily vary depending on the utilized node.js version.

### JSON file
It might be worth considering to introduce a new function that allows to store a stringified version of the parsed dbc file as a json file on the file system. Another new function could then be implemented to parse this json file as an alternative to the existing load method. Parsing json should be really fast. However, the footprint of the json on the file system would probably be much higher.

### Compatibility
As far as I know, the best way to get CAN frames from a CAN interface into node.js is currently this library: https://github.com/sebi2k1/node-can / https://www.npmjs.com/package/socketcan
It can also decode CAN frames, but it does not support dbc files. It might be worth considering a compatibility to this library.
The output object of this library looks like this:
```
{
    "ts_sec": 1669613190,
    "ts_usec": 430472,
    "id": 123,
    "data": [
        1,
        2,
        3,
        4,
        5
    ]
}
```
"ts_sec" is a UNIX timestamp with second precision, "ts_usec" is the remaining microseconds. Both should not be relevant for the decode function.
"id" is a 29 bit integer, the max value is 0x1FFFFFFF (hex) or 536870911 (dec), respectively. So Bit number 29 to 31 are always false. In case the CAN frame contains indeed a 29 bit identifier, then the output object has an additional "ext" property:
```
{
    "ts_sec": 1669613520,
    "ts_usec": 385042,
    "id": 123,
    "ext": true,
    "data": [
        1,
        2,
        3,
        4,
        5
    ]
}
```
The "data" object is a [Buffer](https://nodejs.org/api/buffer.html), **not** an Array.
As far as I can see, the current decode method supports both Buffers and Arrays but the extended flag has to be encoded into the id with Bit number 31 set to true if the CAN message has indeed a 29 bit identifier instead of an 11 bit identifier.

**Comment by bit-dream** [comment]
@nonujigu I went ahead and closed this issue as I released 1.2.0, which contains the new ID map method discussed. On testing I saw roughly a ~45-50% increase in speed of the decode function between the current version and version 1.1.0. I think there are other optimizations we can do to further improve the speed (as you alluded to with your Maps vs Objects comment). I'll be working on a more generic overall speed up of the library, so I think that will be a good place to investigate that.


> I have a few thoughts on this:
> 
> ### Maps vs Objects
> As pointed out in [this article](https://www.zhenghao.io/posts/object-vs-map), utilizing integers as keys for a map could even lead to a performance disadvantage compared to a simple object. So for the decode scenario, a simple object that utilizes stringified integers as keys could even be the best choice. However, it is probably necessary to implement both variants and compare the results to be sure. Also, the results could heavily vary depending on the utilized node.js version.
> 
> ### JSON file
> It might be worth considering to introduce a new function that allows to store a stringified version of the parsed dbc file as a json file on the file system. Another new function could then be implemented to parse this json file as an alternative to the existing load method. Parsing json should be really fast. However, the footprint of the json on the file system would probably be much higher.
> 
> ### Compatibility
> As far as I know, the best way to get CAN frames from a CAN interface into node.js is currently this library: https://github.com/sebi2k1/node-can / https://www.npmjs.com/package/socketcan It can also decode CAN frames, but it does not support dbc files. It might be worth considering a compatibility to this library. The output object of this library looks like this:
> 
> ```
> {
>     "ts_sec": 1669613190,
>     "ts_usec": 430472,
>     "id": 123,
>     "data": [
>         1,
>         2,
>         3,
>         4,
>         5
>     ]
> }
> ```
> 
> "ts_sec" is a UNIX timestamp with second precision, "ts_usec" is the remaining microseconds. Both should not be relevant for the decode function. "id" is a 29 bit integer, the max value is 0x1FFFFFFF (hex) or 536870911 (dec), respectively. So Bit number 29 to 31 are always false. In case the CAN frame contains indeed a 29 bit identifier, then the output object has an additional "ext" property:
> 
> ```
> {
>     "ts_sec": 1669613520,
>     "ts_usec": 385042,
>     "id": 123,
>     "ext": true,
>     "data": [
>         1,
>         2,
>         3,
>         4,
>         5
>     ]
> }
> ```
> 
> The "data" object is a [Buffer](https://nodejs.org/api/buffer.html), **not** an Array. As far as I can see, the current decode method supports both Buffers and Arrays but the extended flag has to be encoded into the id with Bit number 31 set to true if the CAN message has indeed a 29 bit identifier instead of an 11 bit identifier.

I like both the JSON and compability ideas that you brought up and I think those are worth looking into. I'm going to open up a separate issue for both of these.
